File: 0739-daily-temperatures/0739-daily-temperatures.java
class Solution {
    public int[] dailyTemperatures(int[] temperatures) {
        int n = temperatures.length;

        Stack<int[]> st = new Stack<>();

        int[] ans = new int[n];
        for (int i = 0; i < n; i++) {
            while (!st.empty() && st.peek()[0] < temperatures[i]) {
                int idx = st.peek()[1];
                ans[idx] = i - idx;
                st.pop();
            }
            st.push(new int[] { temperatures[i], i });
        }

        return ans;
    }
}

File: 0739-daily-temperatures/0739-daily-temperatures.cpp
class Solution {
public:
    vector<int> dailyTemperatures(vector<int>& temperatures) {

        int n = temperatures.size();
        vector<int> ans(n, 0);

        stack<pair<int, int>> st;

        for (int i = 0; i < n; i++) {
            while (!st.empty() && st.top().first < temperatures[i]) {
                int index = st.top().second;
                ans[index] = i - index;
                st.pop();
            }
            st.push({temperatures[i], i});
        }

        return ans;
    }
};